ccl/changefeedccl: TestChangefeedBasics failed

ccl/changefeedccl.TestChangefeedBasics [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15528860?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/15528860?buildTab=artifacts#/) on master @ [cf1f468e8c04aecd44d614ebe36fe0b1dc49f7e3](https://github.com/cockroachdb/cockroach/commits/cf1f468e8c04aecd44d614ebe36fe0b1dc49f7e3):


```
github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1(0x4003d61520)
	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:1140 +0x1f8
testing.tRunner(0x4003d61520, 0x4001e60f30)
	GOROOT/src/testing/testing.go:1689 +0xec
created by testing.(*T).Run in goroutine 13497
	GOROOT/src/testing/testing.go:1742 +0x318
=== RUN   TestChangefeedBasics/sinkless
    helpers_test.go:867: making server as system tenant
    helpers_test.go:943: making sinkless feed factory
    helpers_test.go:1008: pgURL sinkless SinklessFeedUser
    helpers_test.go:1008: pgURL sinkless root
*
* ERROR: a panic has occurred!
* found 1 short-living non-stopped monitors in in-mem temp storage
* (1) attached stack trace
*   -- stack trace:
*   | runtime.gopanic
*   | 	GOROOT/src/runtime/panic.go:770
*   | [...repeated from below...]
* Wraps: (2) assertion failure
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:657
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).EmergencyStop
*   | 	github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:593
*   | github.com/cockroachdb/cockroach/pkg/server.newSQLServer.func6
*   | 	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:869
*   | github.com/cockroachdb/cockroach/pkg/util/stop.CloserFn.Close
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:108
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop
*   | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:550
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.startTestFullServer.func1
*   | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:424
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.makeSystemServerWithOptions.func1
*   | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:822
*   | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
*   | 	github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/helpers_test.go:1140
*   | testing.tRunner
*   | 	GOROOT/src/testing/testing.go:1689
*   | runtime.goexit
*   | 	src/runtime/asm_arm64.s:1222
* Wraps: (4) found 1 short-living non-stopped monitors in in-mem temp storage
*   | in-mem temp storage 0 B (100 MiB / 0 B)
*   |     flow-disk-monitor 0 B
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
    --- FAIL: TestChangefeedBasics/sinkless (2.77s)
panic: found 1 short-living non-stopped monitors in in-mem temp storage
in-mem temp storage 0 B (100 MiB / 0 B)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedBasics.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39272

--- a/pkg/ccl/changefeedccl/changefeed_dist.go
+++ b/pkg/ccl/changefeedccl/changefeed_dist.go
@@ -315,12 +315,15 @@ func startDistChangefeed(
 
 		jobsprofiler.StorePlanDiagram(ctx, execCfg.DistSQLSrv.Stopper, p, execCfg.InternalDB, jobID)
 
-		// CDC DistSQL flows are long-living, so we want to mark the flow memory
-		// monitor accordingly.
-		planCtx.MarkFlowMonitorAsLongLiving = true
+		// Make sure to use special changefeed monitor going forward as the
+		// parent monitor for the DistSQL infrastructure. This is needed to
+		// prevent a race between the connection that started the changefeed
+		// closing (which closes the current planner's monitor) and changefeed
+		// DistSQL flow being cleaned up.
+		planCtx.OverridePlannerMon = execCfg.DistSQLSrv.ChangefeedMonitor
 		// Copy the evalCtx, as dsp.Run() might change it.
 		evalCtxCopy := *evalCtx
-		// p is the physical plan, recv is the distsqlreceiver
+		// p is the physical plan, recv is the DistSQLReceiver.
 		dsp.Run(ctx, planCtx, noTxn, p, recv, &evalCtxCopy, finishedSetupFn)
 		return resultRows.Err()
 	}

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -183,9 +183,6 @@ func newChangeAggregatorProcessor(
 	}()
 
 	memMonitor := execinfra.NewMonitor(ctx, flowCtx.Mon, "changeagg-mem")
-	// CDC DistSQL flows are long-living, so we mark the memory monitors
-	// accordingly.
-	memMonitor.MarkLongLiving()
 	ca := &changeAggregator{
 		flowCtx:           flowCtx,
 		spec:              spec,
@@ -399,10 +396,7 @@ func (ca *changeAggregator) startKVFeed(
 	opts changefeedbase.StatementOptions,
 ) (kvevent.Reader, chan struct{}, chan error, error) {
 	cfg := ca.flowCtx.Cfg
-	// CDC DistSQL flows are long-living, so we mark the memory monitors
-	// accordingly.
-	const longLiving = true
-	kvFeedMemMon := mon.NewMonitorInheritWithLimit("kvFeed", memLimit, parentMemMon, longLiving)
+	kvFeedMemMon := mon.NewMonitorInheritWithLimit("kvFeed", memLimit, parentMemMon, false /* longLiving */)
 	kvFeedMemMon.StartNoReserved(ctx, parentMemMon)
 	buf := kvevent.NewThrottlingBuffer(
 		kvevent.NewMemBuffer(kvFeedMemMon.MakeBoundAccount(), &cfg.Settings.SV, &ca.metrics.KVFeedMetrics),
@@ -1122,9 +1116,6 @@ func newChangeFrontierProcessor(
 	post *execinfrapb.PostProcessSpec,
 ) (execinfra.Processor, error) {
 	memMonitor := execinfra.NewMonitor(ctx, flowCtx.Mon, "changefntr-mem")
-	// CDC DistSQL flows are long-living, so we mark the memory monitors
-	// accordingly.
-	memMonitor.MarkLongLiving()
 	sf, err := makeSchemaChangeFrontier(hlc.Timestamp{}, spec.TrackedSpans...)
 	if err != nil {
 		return nil, err

--- a/pkg/ccl/changefeedccl/metrics.go
+++ b/pkg/ccl/changefeedccl/metrics.go
@@ -1136,6 +1136,42 @@ func MakeMetrics(histogramWindow time.Duration) metric.Struct {
 	return m
 }
 
+var (
+	metaMemMaxBytes = metric.Metadata{
+		Name:        "sql.mem.changefeed.max",
+		Help:        "Maximum memory usage across all changefeeds",
+		Measurement: "Memory",
+		Unit:        metric.Unit_BYTES,
+	}
+	metaMemCurBytes = metric.Metadata{
+		Name:        "sql.mem.changefeed.current",
+		Help:        "Current memory usage across all changefeeds",
+		Measurement: "Memory",
+		Unit:        metric.Unit_BYTES,
+	}
+)
+
+// See pkg/sql/mem_metrics.go
+// log10int64times1000 = log10(math.MaxInt64) * 1000, rounded up somewhat
+const log10int64times1000 = 19 * 1000
+
+// MakeMemoryMetrics instantiates the metrics holder for memory monitors of
+// changefeeds.
+func MakeMemoryMetrics(
+	histogramWindow time.Duration,
+) (curCount *metric.Gauge, maxHist metric.IHistogram) {
+	curCount = metric.NewGauge(metaMemCurBytes)
+	maxHist = metric.NewHistogram(metric.HistogramOptions{
+		Metadata:     metaMemMaxBytes,
+		Duration:     histogramWindow,
+		MaxVal:       log10int64times1000,
+		SigFigs:      3,
+		BucketConfig: metric.MemoryUsage64MBBuckets,
+	})
+	return curCount, maxHist
+}
+
 func init() {
 	jobs.MakeChangefeedMetricsHook = MakeMetrics
+	jobs.MakeChangefeedMemoryMetricsHook = MakeMemoryMetrics
 }

--- a/pkg/jobs/metrics.go
+++ b/pkg/jobs/metrics.go
@@ -297,6 +297,10 @@ func (m *Metrics) init(histogramWindowInterval time.Duration) {
 // ccl code.
 var MakeChangefeedMetricsHook func(time.Duration) metric.Struct
 
+// MakeChangefeedMemoryMetricsHook allows for registration of changefeed memory
+// metrics from ccl code.
+var MakeChangefeedMemoryMetricsHook func(time.Duration) (curCount *metric.Gauge, maxHist metric.IHistogram)
+
 // MakeStreamIngestMetricsHook allows for registration of streaming metrics from
 // ccl code.
 var MakeStreamIngestMetricsHook func(duration time.Duration) metric.Struct

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -1024,6 +1024,7 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 			cfg.TableStatCacheSize,
 			cfg.Settings,
 			cfg.internalDB,
+			cfg.stopper,
 		),
 
 		QueryCache:                 querycache.New(cfg.QueryCacheSize),

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -696,17 +696,26 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 	bulkMetrics := bulk.MakeBulkMetrics(cfg.HistogramWindowInterval())
 	cfg.registry.AddMetricStruct(bulkMetrics)
 	bulkMemoryMonitor.SetMetrics(bulkMetrics.CurBytesCount, bulkMetrics.MaxBytesHist)
-	bulkMemoryMonitor.StartNoReserved(context.Background(), rootSQLMemoryMonitor)
+	bulkMemoryMonitor.StartNoReserved(ctx, rootSQLMemoryMonitor)
 
 	backfillMemoryMonitor := execinfra.NewMonitor(ctx, bulkMemoryMonitor, "backfill-mon")
 	backfillMemoryMonitor.MarkLongLiving()
 	backupMemoryMonitor := execinfra.NewMonitor(ctx, bulkMemoryMonitor, "backup-mon")
 	backupMemoryMonitor.MarkLongLiving()
 
+	changefeedMemoryMonitor := mon.NewMonitorInheritWithLimit(
+		"changefeed-mon", 0 /* limit */, rootSQLMemoryMonitor, true, /* longLiving */
+	)
+	if jobs.MakeChangefeedMemoryMetricsHook != nil {
+		changefeedCurCount, changefeedMaxHist := jobs.MakeChangefeedMemoryMetricsHook(cfg.HistogramWindowInterval())
+		changefeedMemoryMonitor.SetMetrics(changefeedCurCount, changefeedMaxHist)
+	}
+	changefeedMemoryMonitor.StartNoReserved(ctx, rootSQLMemoryMonitor)
+
 	serverCacheMemoryMonitor := mon.NewMonitorInheritWithLimit(
 		"server-cache-mon", 0 /* limit */, rootSQLMemoryMonitor, true, /* longLiving */
 	)
-	serverCacheMemoryMonitor.StartNoReserved(context.Background(), rootSQLMemoryMonitor)
+	serverCacheMemoryMonitor.StartNoReserved(ctx, rootSQLMemoryMonitor)
 
 	// Set up the DistSQL temp engine.
 
@@ -821,6 +830,7 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 		ParentDiskMonitor: cfg.TempStorageConfig.Mon,
 		BackfillerMonitor: backfillMemoryMonitor,
 		BackupMonitor:     backupMemoryMonitor,
+		ChangefeedMonitor: changefeedMemoryMonitor,
 		BulkSenderLimiter: bulkSenderLimiter,
 
 		ParentMemoryMonitor: rootSQLMemoryMonitor,

--- a/pkg/sql/distsql/server.go
+++ b/pkg/sql/distsql/server.go
@@ -184,15 +184,12 @@ func (ds *ServerImpl) setDraining(drain bool) error {
 
 // setupFlow creates a Flow.
 //
-// Args:
-// reserved: Specifies the upfront memory reservation that the flow takes
+//   - reserved: specifies the upfront memory reservation that the flow takes
+//     ownership of. This account is already closed if an error is returned or
+//     will be closed through Flow.Cleanup.
 //
-//	ownership of. This account is already closed if an error is returned or
-//	will be closed through Flow.Cleanup.
-//
-// localState: Specifies if the flow runs entirely on this node and, if it does,
-//
-//	specifies the txn and other attributes.
+//   - localState: specifies if the flow runs entirely on this node and, if it
+//     does, specifies the txn and other attributes.
 //
 // Note: unless an error is returned, the returned context contains a span that
 // must be finished through Flow.Cleanup.
@@ -263,11 +260,10 @@ func (ds *ServerImpl) setupFlow(
 	}
 
 	monitor = mon.NewMonitor(mon.Options{
-		Name:       "flow " + redact.RedactableString(req.Flow.FlowID.Short()),
-		CurCount:   ds.Metrics.CurBytesCount,
-		MaxHist:    ds.Metrics.MaxBytesHist,
-		Settings:   ds.Settings,
-		LongLiving: localState.MarkFlowMonitorAsLongLiving,
+		Name:     "flow " + redact.RedactableString(req.Flow.FlowID.Short()),
+		CurCount: ds.Metrics.CurBytesCount,
+		MaxHist:  ds.Metrics.MaxBytesHist,
+		Settings: ds.Settings,
 	})
 	monitor.Start(ctx, parentMonitor, reserved)
 	diskMonitor = execinfra.NewMonitor(ctx, ds.ParentDiskMonitor, "flow-disk-monitor")
@@ -560,11 +556,6 @@ type LocalState struct {
 	// mapping to coldata.Batch, use any to avoid injecting new
 	// dependencies.
 	LocalVectorSources map[int32]any
-
-	// MarkFlowMonitorAsLongLiving, if set, instructs the DistSQL runner to mark
-	// the "flow" memory monitor as long-living one, thus exempting it from
-	// having to be stopped when the txn monitor is stopped.
-	MarkFlowMonitorAsLongLiving bool
 }
 
 // MustUseLeafTxn returns true if a LeafTxn must be used. It is valid to call

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -59,6 +59,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/intsets"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
@@ -919,10 +920,9 @@ type PlanningCtx struct {
 	// This is true if plan is a simple insert that can be vectorized.
 	isVectorInsert bool
 
-	// MarkFlowMonitorAsLongLiving, if set, instructs the DistSQL runner to mark
-	// the "flow" memory monitor as long-living one, thus exempting it from
-	// having to be stopped when the txn monitor is stopped.
-	MarkFlowMonitorAsLongLiving bool
+	// OverridePlannerMon, if set, will be used instead of the Planner.Mon() as
+	// the parent monitor for the DistSQL flow.
+	OverridePlannerMon *mon.BytesMonitor
 }
 
 var _ physicalplan.ExprContext = &PlanningCtx{}

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -466,7 +466,11 @@ func (dsp *DistSQLPlanner) setupFlows(
 		batchReceiver = recv
 	}
 	origCtx := ctx
-	ctx, flow, opChains, err := dsp.distSQLSrv.SetupLocalSyncFlow(ctx, evalCtx.Planner.Mon(), &setupReq, recv, batchReceiver, localState)
+	parentMonitor := evalCtx.Planner.Mon()
+	if planCtx.OverridePlannerMon != nil {
+		parentMonitor = planCtx.OverridePlannerMon
+	}
+	ctx, flow, opChains, err := dsp.distSQLSrv.SetupLocalSyncFlow(ctx, parentMonitor, &setupReq, recv, batchReceiver, localState)
 	if err == nil && planCtx.saveFlows != nil {
 		err = planCtx.saveFlows(flows, opChains, planCtx.infra.LocalProcessors, isVectorized)
 	}
@@ -696,7 +700,6 @@ func (dsp *DistSQLPlanner) Run(
 	localState.Txn = txn
 	localState.LocalProcs = plan.LocalProcessors
 	localState.LocalVectorSources = plan.LocalVectorSources
-	localState.MarkFlowMonitorAsLongLiving = planCtx.MarkFlowMonitorAsLongLiving
 	if planCtx.planner != nil {
 		// Note that the planner's collection will only be used for local plans.
 		localState.Collection = planCtx.planner.Descriptors()

--- a/pkg/sql/execinfra/server_config.go
+++ b/pkg/sql/execinfra/server_config.go
@@ -116,6 +116,9 @@ type ServerConfig struct {
 	// used during restore.
 	RestoreMonitor *mon.BytesMonitor
 
+	// ChangefeedMonitor is the parent monitor for all CDC DistSQL flows.
+	ChangefeedMonitor *mon.BytesMonitor
+
 	// BulkSenderLimiter is the concurrency limiter that is shared across all of
 	// the processes in a given sql server when sending bulk ingest (AddSST) reqs.
 	BulkSenderLimiter limit.ConcurrentRequestLimiter

--- a/pkg/sql/plan.go
+++ b/pkg/sql/plan.go
@@ -546,7 +546,7 @@ func (p *planner) maybePlanHook(ctx context.Context, stmt tree.Statement) (planN
 					"cannot execute prepared %v statement",
 					planHook.name,
 				)
-			}, header, nil), nil
+			}, header, nil /* subplans */, p.execCfg.DistSQLPlanner.stopper), nil
 		}
 
 		if fn, header, subplans, avoidBuffering, err := planHook.fn(ctx, stmt, p); err != nil {
@@ -555,7 +555,7 @@ func (p *planner) maybePlanHook(ctx context.Context, stmt tree.Statement) (planN
 			if avoidBuffering {
 				p.curPlan.avoidBuffering = true
 			}
-			return newHookFnNode(planHook.name, fn, header, subplans), nil
+			return newHookFnNode(planHook.name, fn, header, subplans, p.execCfg.DistSQLPlanner.stopper), nil
 		}
 	}
 	return nil, nil

--- a/pkg/sql/planhook.go
+++ b/pkg/sql/planhook.go
@@ -33,7 +33,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/upgrade"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
 )
 
 // planHookFn is a function that can intercept a statement being planned and
@@ -158,6 +158,7 @@ type hookFnNode struct {
 	f        PlanHookRowFn
 	header   colinfo.ResultColumns
 	subplans []planNode
+	stopper  *stop.Stopper
 
 	run hookFnRun
 }
@@ -173,40 +174,37 @@ type hookFnRun struct {
 }
 
 func newHookFnNode(
-	name string, fn PlanHookRowFn, header colinfo.ResultColumns, subplans []planNode,
+	name string,
+	fn PlanHookRowFn,
+	header colinfo.ResultColumns,
+	subplans []planNode,
+	stopper *stop.Stopper,
 ) *hookFnNode {
-	return &hookFnNode{name: name, f: fn, header: header, subplans: subplans}
+	return &hookFnNode{name: name, f: fn, header: header, subplans: subplans, stopper: stopper}
 }
 
 func (f *hookFnNode) startExec(params runParams) error {
-	// TODO(dan): Make sure the resultCollector is set to flush after every row.
 	f.run.resultsCh = make(chan tree.Datums)
 	f.run.errCh = make(chan error)
-	// Start a new span for the execution of the hook's plan. This is particularly
-	// important since that execution might outlive the span in params.ctx.
-	// Generally speaking, the subplan is not supposed to outlive the caller since
-	// hookFnNode.Next() is supposed to be called until the subplan is exhausted.
-	// However, there's no strict protocol in place about the goroutines that the
-	// subplan might spawn. For example, if the subplan creates a DistSQL flow,
-	// the cleanup of that flow might race with an error bubbling up to Next(). In
-	// particular, there seem to be races around context cancelation, as Next()
-	// listens for cancellation for better or worse.
-	//
-	// TODO(andrei): We should implement a protocol where the hookFnNode doesn't
-	// listen for cancellation and guarantee Next() doesn't return false until the
-	// subplan has completely shutdown.
-	subplanCtx, sp := tracing.ChildSpan(params.ctx, f.name)
-	go func() {
-		defer sp.Finish()
-		err := f.f(subplanCtx, f.subplans, f.run.resultsCh)
-		select {
-		case <-params.ctx.Done():
-		case f.run.errCh <- err:
-		}
-		close(f.run.errCh)
-		close(f.run.resultsCh)
-	}()
-	return nil
+	// Note that it's ok if the async task is not started due to server shutdown
+	// because the context should be canceled then too, which would unblock
+	// calls to Next if they happen.
+	return f.stopper.RunAsyncTaskEx(
+		params.ctx,
+		stop.TaskOpts{
+			TaskName: f.name,
+			SpanOpt:  stop.ChildSpan,
+		},
+		func(ctx context.Context) {
+			err := f.f(ctx, f.subplans, f.run.resultsCh)
+			select {
+			case <-ctx.Done():
+			case f.run.errCh <- err:
+			}
+			close(f.run.errCh)
+			close(f.run.resultsCh)
+		},
+	)
 }
 
 func (f *hookFnNode) Next(params runParams) (bool, error) {

--- a/pkg/sql/stats/BUILD.bazel
+++ b/pkg/sql/stats/BUILD.bazel
@@ -61,7 +61,6 @@ go_library(
         "//pkg/util/syncutil",
         "//pkg/util/timeutil",
         "//pkg/util/timeutil/pgdate",
-        "//pkg/util/tracing",
         "@com_github_cockroachdb_errors//:errors",
         "@com_github_cockroachdb_redact//:redact",
     ],

--- a/pkg/sql/stats/automatic_stats_test.go
+++ b/pkg/sql/stats/automatic_stats_test.go
@@ -71,6 +71,7 @@ func TestMaybeRefreshStats(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
@@ -195,6 +196,7 @@ func TestEnsureAllTablesQueries(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 	r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
@@ -297,6 +299,7 @@ func BenchmarkEnsureAllTables(b *testing.B) {
 				10, /* cacheSize */
 				s.ClusterSettings(),
 				s.InternalDB().(descs.DB),
+				s.AppStopper(),
 			)
 			require.NoError(b, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 			r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
@@ -370,6 +373,7 @@ func TestAverageRefreshTime(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
@@ -617,6 +621,7 @@ func TestAutoStatsReadOnlyTables(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
@@ -672,6 +677,7 @@ func TestAutoStatsOnStartupClusterSettingOff(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 	refresher := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
@@ -719,6 +725,7 @@ func TestNoRetryOnFailure(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 	r := MakeRefresher(s.AmbientCtx(), st, internalDB, cache, time.Microsecond /* asOfTime */, nil /* knobs */)
@@ -836,6 +843,7 @@ func TestAnalyzeSystemTables(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, codec, s.RangeFeedFactory().(*rangefeed.Factory)))
 

--- a/pkg/sql/stats/delete_stats_test.go
+++ b/pkg/sql/stats/delete_stats_test.go
@@ -48,6 +48,7 @@ func TestDeleteOldStatsForColumns(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		db,
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, s.Codec(), s.RangeFeedFactory().(*rangefeed.Factory)))
 
@@ -348,6 +349,7 @@ func TestDeleteOldStatsForOtherColumns(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, cache.Start(ctx, s.Codec(), s.RangeFeedFactory().(*rangefeed.Factory)))
 	testData := []TableStatisticProto{

--- a/pkg/sql/stats/stats_cache.go
+++ b/pkg/sql/stats/stats_cache.go
@@ -37,8 +37,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
 )
 
@@ -72,6 +72,7 @@ type TableStatisticsCache struct {
 	}
 	db       descs.DB
 	settings *cluster.Settings
+	stopper  *stop.Stopper
 
 	// Used when decoding KV from the range feed.
 	datumAlloc tree.DatumAlloc
@@ -120,11 +121,12 @@ type cacheEntry struct {
 // NewTableStatisticsCache creates a new TableStatisticsCache that can hold
 // statistics for <cacheSize> tables.
 func NewTableStatisticsCache(
-	cacheSize int, settings *cluster.Settings, db descs.DB,
+	cacheSize int, settings *cluster.Settings, db descs.DB, stopper *stop.Stopper,
 ) *TableStatisticsCache {
 	tableStatsCache := &TableStatisticsCache{
 		db:       db,
 		settings: settings,
+		stopper:  stopper,
 	}
 	tableStatsCache.mu.cache = cache.NewUnorderedCache(cache.Config{
 		Policy:      cache.CacheLRU,
@@ -520,12 +522,11 @@ func (sc *TableStatisticsCache) refreshTableStats(
 	ctx context.Context, tableID descpb.ID, ts hlc.Timestamp,
 ) {
 	log.VEventf(ctx, 1, "refreshing statistics for table %d", tableID)
-	ctx, span := tracing.ForkSpan(ctx, "refresh-table-stats")
-	// Perform an asynchronous refresh of the cache.
-	go func() {
-		defer span.Finish()
+	// Perform an asynchronous refresh of the cache. An error is returned only
+	// on the server shutdown at which point we don't care about the refresh.
+	_ = sc.stopper.RunAsyncTask(ctx, "refresh-table-stats", func(ctx context.Context) {
 		sc.refreshCacheEntry(ctx, tableID, ts)
-	}()
+	})
 }
 
 // InvalidateTableStats invalidates the cached statistics for the given table ID.

--- a/pkg/sql/stats/stats_cache_test.go
+++ b/pkg/sql/stats/stats_cache_test.go
@@ -236,7 +236,7 @@ func TestCacheBasic(t *testing.T) {
 	// Create a cache and iteratively query the cache for each tableID. This
 	// will result in the cache getting populated. When the stats cache size is
 	// exceeded, entries should be evicted according to the LRU policy.
-	sc := NewTableStatisticsCache(2 /* cacheSize */, s.ClusterSettings(), db)
+	sc := NewTableStatisticsCache(2 /* cacheSize */, s.ClusterSettings(), db, s.AppStopper())
 	require.NoError(t, sc.Start(ctx, s.Codec(), s.RangeFeedFactory().(*rangefeed.Factory)))
 	for _, tableID := range tableIDs {
 		checkStatsForTable(ctx, t, sc, expectedStats[tableID], tableID)
@@ -338,7 +338,7 @@ func TestCacheUserDefinedTypes(t *testing.T) {
 	insqlDB := s.InternalDB().(descs.DB)
 
 	// Make a stats cache.
-	sc := NewTableStatisticsCache(1, s.ClusterSettings(), insqlDB)
+	sc := NewTableStatisticsCache(1, s.ClusterSettings(), insqlDB, s.AppStopper())
 	require.NoError(t, sc.Start(ctx, s.Codec(), s.RangeFeedFactory().(*rangefeed.Factory)))
 	tbl := desctestutils.TestingGetPublicTableDescriptor(kvDB, s.Codec(), "t", "tt")
 	// Get stats for our table. We are ensuring here that the access to the stats
@@ -391,7 +391,7 @@ func TestCacheWait(t *testing.T) {
 		tableIDs = append(tableIDs, tableID)
 	}
 	sort.Sort(tableIDs)
-	sc := NewTableStatisticsCache(len(tableIDs) /* cacheSize */, s.ClusterSettings(), db)
+	sc := NewTableStatisticsCache(len(tableIDs) /* cacheSize */, s.ClusterSettings(), db, s.AppStopper())
 	require.NoError(t, sc.Start(ctx, s.Codec(), s.RangeFeedFactory().(*rangefeed.Factory)))
 	for _, tableID := range tableIDs {
 		checkStatsForTable(ctx, t, sc, expectedStats[tableID], tableID)
@@ -443,6 +443,7 @@ func TestCacheAutoRefresh(t *testing.T) {
 		10, /* cacheSize */
 		s.ClusterSettings(),
 		s.InternalDB().(descs.DB),
+		s.AppStopper(),
 	)
 	require.NoError(t, sc.Start(ctx, s.Codec(), s.RangeFeedFactory().(*rangefeed.Factory)))
 

